### ISSUE TITLE
Add functionality to download compiler toolchain

### ISSUE THREAD

**Issue by seantalts** [stan-dev/cmdstanpy#7]
It would be awesome to provide a scripted shortcut in the cmdstanpy library that either helps a user install a compiler toolchain or downloads one for them to some specified directory that we then use. Ideally the latter so that we could help users without administrative privileges (especially common on Windows). 

R has a package that attempts to perform the global version: https://github.com/r-lib/pkgbuild. I wonder if something similar exists in the Python world that we can adapt.

**Comment by maedoc** [comment]
Python users on the Anaconda distributions can install toolchains with the `conda` package manager.  CmdStan could be packaged as another conda package, then a user could just.  Conda handles dependencies, so in principle it could just be `conda install cmdstanpy=2.18.1`.

**Comment by seantalts** [comment]
Hmm, are there users who wouldn't be able to use conda for some reason?

**Comment by maedoc** [comment]
Not *everyone* uses it, but it is available on mac, linux and windows.

**Comment by ahartikainen** [comment]
What if we download clang/mingw-w64 to local directory under cmdstan folder, and then "hardlink" compiler to that. This would not need any setup in python side.

**Comment by maedoc** [comment]
> we manage to use clang + libc++ on Windows

it'd be nice to have that everywhere, not just on Windows.  The LLVM project builds Clang for multiple platforms, but not libc++, which require CMake to build.

It'd be neat to have a simple Python API for making a C++ toolchain available, but it seems out of scope for this library.

**Comment by seantalts** [comment]
> What if we download clang/mingw-w64 to local directory under cmdstan folder, and then "hardlink" compiler to that. This would not need any setup in python side.

> Even better if we manage to use clang + libc++ on Windows.

I think this would work. Not sure exactly what you mean by "hardlink" but worst case we can prepend those directories to the PATH whenever we execute a command.

> It'd be neat to have a simple Python API for making a C++ toolchain available, but it seems out of scope for this library.

Figuring out how to automatically get a C++ toolchain in place is a core part of our vision to make installation seamless for users. That said, it doesn't have to be organized such that it's part of this CmdStanPy pypi package or even github repo, and maybe it would be better separated out (as long as we can integrate it seamlessly for users).

**Comment by ahartikainen** [comment]
Hardlink meaning --> add absolutpath to compiler somewhere.

Just one question... how hard would it be to write that makefile in python? Because at the end we want call gcc/clang +stuff.

**Comment by ahartikainen** [comment]
I'm thinking Stan `makefile`. Currently Windows users need `make` + `c++ compiler`.

If we can cut the `make` (or provide local one) then implementing C++ compiler might be easier.

I don't know if we could download these from conda (without conda)

    https://anaconda.org/conda-forge/libcxx
    https://anaconda.org/conda-forge/clangdev

Then linking stuff should be more or less static right?

**Comment by maedoc** [comment]
Conda-forge has a `make` too, but `distutils` ships an [API for compilers](https://docs.python.org/2/distutils/apiref.html#module-distutils.ccompiler) which could be nicer to work with, if deviations from the CmdStan makefiles are required.

**Comment by ahartikainen** [comment]
hmm, true, but then this package has same burden as pystan. E.g. all the problems with distutils being slow to update.

So let's keep makefile.

**Comment by syclik** [comment]
The makefile isn't necessary. We could call the compiler directly.

Make is useful for building when dependencies are updated. If we don't need
that, we can just invoke stanc, then the c++ compiler, and finally the
linker to compile an executable.

It starts getting messy taking that approach when you start having to
branch for single threaded, multithreaded, or MPI, or some other options,
but it's not that bad.

**Comment by ahartikainen** [comment]
Why would we need different branches if that functionality is in python functions?

In the end we are calling commandline.

**Comment by maedoc** [comment]
a lot of expertise on compilers options (optimization, threading libs, cross platform goodness) is tied up in those makefiles, it makes sense to reuse that (unless/until there's something like `llvm-config` for Stan to generate those options w/o parsing Makefiles)

**Comment by syclik** [comment]
> Why would we need different branches if that functionality is in python functions?

@ahartikainen, by "branch" I meant conditional branches, not git branches.

> a lot of expertise on compilers options (optimization, threading libs, cross platform goodness) is tied up in those makefiles, it makes sense to reuse that (unless/until there's something like llvm-config for Stan to generate those options w/o parsing Makefiles)

Yes, but the most complicated parts of it deal with dependency management. The build is trivial and should really be three calls (the three I outlined). If we assumed that things inside CmdStan don't change, then we really don't need that heavy makefile to build. But, as you mentioned, some process has to deal with optimization, threading libs, cross platform goodness, etc. (No free lunch, unfortunately.)

**Comment by maedoc** [comment]
> with conda from pip install conda? 

That's a good question: I'd guess pip users can install conda, create a dedicated cmdstan toolchain environment as above, and use conda+env solely for compilation purposes.  In other words, the environment the PoC above creates, is independent on the conda env or pip env used for your project code.  It could be done in the same env, but I'm always a little worried about dependency versions conflicting later down the road. In one's project code, the `cmdstanpy` module could check for a folder `~/.conda/envs/cmdstan-toolchain` or similar. 

> process should probably work on Windows too... with some small changes

the above script rewritten in Python would be portable, but I imagine flags will be different.  On macOS I need to include Brew-installed GCC headers for the likes of `unistd.h`.. not sure why.

**Comment by seantalts** [comment]
Does it matter that I don't have the `MacOSX10.14.sdk` directory on my Mac? Is that just to fix a bug in 10.14?

If all Mac takes to work is `xcode-select --install` that would be great, we could just offer to run that for them, but it seems like that might not work on the latest Mac OS X?

**Comment by ahartikainen** [comment]
Just a question. Why do we need xcode if we have clang + libcxx?

**Comment by maedoc** [comment]
> Does it matter that I don't have the MacOSX10.14.sdk directory on my Mac?

I ran `clang++ -v### -xc++ /dev/null` to see what command the macOS' clang driver is actually invoking, so the output of that on your Mac, assuming you have some valid Xcode, should provide the right folder.

> Why do we need xcode if we have clang + libcxx?

LLVM doesn't ship a libc, and C++ is a superset of C. Since libc is always present, I'm pretty sure just the headers are required.  I'd guess a per-platform set could just be copied into the LLVM include folder and it'd work.

**Comment by maedoc** [comment]
> do we need xcode if we have clang + libcxx?

to answer the slightly different question you didn't ask, maybe not, and that'd be neat, considering it eats a ton of space on a precious non upgradeable SSD

**Comment by seantalts** [comment]
Yeah, we may not need it - was just thinking installing the xcode command line tools with `xcode-select --install` could have been an easy path towards the end goal of ensuring a user has a C++ toolchain. But I also thought of another reason that's not as good - the Apple Xcode clang doesn't (or didn't?) support threads (that may be fixed now).

**Comment by maedoc** [comment]
maybe I'm wrong but I think one needs the full Xcode for a complete C++ toolchain.  At least on Mojave, I tried this (i.e. just the command line tools) and couldn't use make/clang

**Comment by seantalts** [comment]
What exactly did you try? There is a specific thing called the "xcode command line tools" that is installed with `xcode-select --install` that is what I was talking about (see [here](https://apple.stackexchange.com/questions/337744/installing-xcode-command-line-tools)).  If that doesn't install make and clang anymore in Mojave that's really sad (e.g. what tools DID they install then? haha).

But maybe all of that is irrelevant if we figure out a way to package clang ourselves...

**Comment by maedoc** [comment]
I ran into the same confusion as in the poster in your link.  Maybe I should've just tried a bit longer.  The real test of using one's own Clang is if it works on a fresh system without Xcode or Xcode CLT.

I gave it a go on Windows as well, but it's not clear that the GCC-compatible Clang driver produces correct linkage, and at the last linking step, there's some funny business about vcruntime.dll.  A CMake: build for Cmdstan would be a workaround, using the MSVC CL-compatible Clang wrapper to get all the LLVM goodness.

**Comment by ahartikainen** [comment]
Installing msvc (e.g. 2017).

Clang-cl uses msvc headers right?

**Comment by ahartikainen** [comment]
I think we should start with a function to download and install RTools 4 (or 3.5) to specific dir (`~/.cmdstanpy` by default) and then add option to edit make/local or env variables for `CC`, `CXX` and `MAKE`. make could be checked and changed at import time, if it is found (from rtools) and not defined before.

Then we could follow install_cmdstan format and create a script for it. Also install_cmdstan could call it if needed.

**Comment by mitzimorris** [comment]
command-line function (Windows/RTools only) `install_cxx_toolchain` seems to be working quite nicely.
question for @ahartikainen:  does this need to be run as a command line script, or could we make it a function, similar to how `install_cmdstan` works?

**Comment by seantalts** [comment]
Woo! Thank you!
